doc/benchexec-in-container.md

```diff
@@ -0,0 +1,233 @@
+instead.)
+
+In case you want to modify this setup, please consider the notes mentioned
+[below](#background-and-technical-details).
+
+### Securing Docker Execution
+
+When running under Docker, the user which executes the commands by default has
+`UID` 0, i.e. has the permissions of `root` on the host system. For example, if
+you mount a folder into the container, the processes started in the container
+have *full write access* to these files, no matter what their permissions are.
+As such, a small mistake in, e.g., an evaluation script that deletes temporary
+files, could easily wreak havoc on your system. Using Podman directly mitigates
+this issue, as the Podman container runs with the permissions of the current
+user, not `root`, and thus cannot mess with system files (in particular, the
+"`root`" user in a Podman container is "mapped" to your current user).
+
+To significantly increase security of your Docker execution, add
+```docker
+# Create non-root user and set it as default
+RUN useradd -ms /bin/bash user
+WORKDIR /home/user
+USER user
+```
+at the end of your Dockerfile. Note: This is still worse than using Podman
+(e.g. privilege escalation within the container leads to obtaining root
+permissions on the host) but drastically reduces the chance for accidental
+problems.
+
+### Non-interactive Setups
+
+In case you intend to only execute one BenchExec process inside the container,
+the setup simplifies notably. In particular, then, it is sufficient to execute
+the BenchExec process as entry point, without requiring the `init.sh` script.
+For example, if BenchExec is installed in a container with tag `my-benchexec`,
+you can simply execute
+`podman run <same flags as above> -t my-benchexec benchexec <arguments>`
+(analogous with Docker).
+
+This is mostly useful for executing a large benchmark suite with `benchexec`
+inside a container with the benchmarked tools installed, but processing the
+results separately.
+
+## Background and Technical Details
+
+Just like BenchExec, tools like Docker and Podman make heavy use of cgroups to
+isolate processes. Thus, in order to get them working nicely together, some
+additional tricks are required, especially in an interactive container (e.g.
+one with a shell, where multiple commands can be executed).
+
+The main complication of using BenchExec with Docker is that benchexec needs a
+"clean" cgroup together with enabled controllers. On a standard system, cgroups
+are usually managed by, for example, systemd, from which BenchExec can
+transparently obtain its cgroup. In contrast, in an interactive container, the
+root process usually is just a shell, which is placed in the "root" cgroup of
+the container. Due to the "no internal processes" rule (see the
+[cgroup documentation](https://man.archlinux.org/man/cgroups.7) for more
+information) any (non-root) cgroup can *either* contain processes *or* delegate
+controllers to a child group. Thus, the shell in the root cgroup of the
```

Comment:
```suggestion
the container. Due to the "no internal processes" rule (see the
[cgroup documentation](https://www.man7.org/linux/man-pages/man7/cgroups.7.html)
for more information)
a cgroup can *either* contain processes *or* delegate controllers to a child group
(the exception for the root cgroup only applies for the root cgroup of the host).
Thus, the shell in the root cgroup of the
```
We use the official links to the man-page project. And I think we can shorten the discussion of the special root case here.

Reply:
ah, I wasn't sure which page is the most "official" one :D